src/scene/materials/shader-graph-builder.js

```diff
@@ -0,0 +1,128 @@
+import { ShaderGraphNode } from './shader-graph-node.js';
+
+var id = 0;
+
+/**
+ * @private
+ * @class
+ * @name pc.ShaderGraphBuilder
+ * @classdesc A Shader Graph Builder class
+ * @param {pc.Application} app - Application with shader graph core nodes registered that will be used to build the shader graph
+ */
+var ShaderGraphBuilder = function (app) {
+    id++;
+
+    this.app = app;
+    this._nodes = this.app.shaderNodes._nodes;
+
+    this._graph = new ShaderGraphNode();
+    this._graph.name = 'graphRoot_' + id;
+};
+
+ShaderGraphBuilder.prototype.constructor = ShaderGraphBuilder;
+
+Object.assign(ShaderGraphBuilder.prototype, {
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#addParam
+     * @description adds a parameter input to graph
+     * @param {string} type - type of parameter
+     * @param {string} name - name of parameter
+     * @param {any} value - optional value of parameter
+     * @returns {any} returns the created input port
+     */
+    addParam: function (type, name, value) {
+        var ioPort = this._graph.addInput(type, name, value);

```

Comment:
handle and print error when .addInput fails (duplicates, incorrect type ...)

Reply:
definitely after MVP

Reply:
added validation code to the builder class